cmd/channel/distributed/dispatcher/main.go

```diff
@@ -219,21 +210,35 @@ func flush(logger *zap.Logger) {
 	eventingmetrics.FlushExporter()
 }
 
-// secretObserver is the callback function that handles changes to our Secret
-func secretObserver(ctx context.Context, secret *corev1.Secret) {
-	secretLogger := logging.FromContext(ctx)
-
-	if secret == nil {
-		secretLogger.Warn("Nil Secret passed to secretObserver; ignoring")
-		return
+// NewSecretObserver is a factory for creating the callback function that handles changes to the Kafka Secret.
+func NewSecretObserver(kcController *kncontroller.Impl, channelKey string, dispatcher dispatch.Dispatcher) func(ctx context.Context, secret *corev1.Secret) {
+	return func(ctx context.Context, secret *corev1.Secret) {
+
+		// Get The Logger From The Context
+		logger := logging.FromContext(ctx)
+
+		// Validate The Secret (Ignore Invalid)
+		if secret == nil {
+			logger.Warn("Nil Secret passed to secretObserver; ignoring")
+			return
+		}
+
+		// Ignore Startup Scenario Where Dispatcher Reference Might Be Nil
+		if dispatcher == nil {
+			logger.Debug("Dispatcher is nil during call to secretObserver; ignoring changes")
+			return
+		}
+
+		// Signal The Dispatcher To Recreate Kafka Config And Reconnect All Current ConsumerGroups
+		dispatcher.SecretChanged(ctx, secret)
+
+		// Requeue The KafkaChannel To Fix Any "Stuck" Subscriptions From A Prior Bad Kafka Secret
+		channelNamespace, channelName, err := cache.SplitMetaNamespaceKey(channelKey)
+		if err != nil {
+			logger.Error("Invalid KafkaChannel Key", zap.String("ChannelKey", channelKey), zap.Error(err))
+		} else {
+			logger.Debug("Requeue-ing KafkaChannel due to Kafka Secret change", zap.String("ChannelKey", channelKey))
+			kcController.EnqueueKey(types.NamespacedName{Namespace: channelNamespace, Name: channelName})
+		}
```

Comment:
Instead of passing in the channelKey and doing this SplitMetaNamespaceKey every time (with the error handling), would it be better to just pass in the NamespacedName to NewSecretObserver?  If I'm not mistaken it will never change during the life of the pod and would make this function less complicated.

Reply:
You're right that the `channelKey` value is "fixed" for the Dispatcher instance (since it's only ever dealing with one KafkaChannel) but I'd rather not pass in the NamespacedName since it's only used here, etc...